Publishing typescript library code

GoodMorningA1i opened the pull request.

mieubrisse commented on typescript/package.json:
```diff
@@ -8,5 +8,32 @@
+    "kurtosis-js",
+    "lib"
+  ],
+  "author": "Kurtosis <accounts@kurtosistech.com> (https://www.kurtosistech.com/)",
```
You'll want this to be Kurtosis Technologies Inc

mieubrisse commented on typescript/package.json:
```diff
@@ -8,5 +8,32 @@
+  "name": "kurtosis-js-lib",
+  "version": "1.0.0",
+  "description": "This repo contains clients for communicating with the Kurtosis API server, which both testsuites and Kurtosis Modules will use to manipulate the test environment.",
+  "main": "build/index.js",
```
Does this work? The `index.ts` compiles to `output.js`, no?

mieubrisse commented on typescript/package.json:
```diff
@@ -8,5 +8,32 @@
     "@types/mocha": "^9.0.0"
+  },
+  "name": "kurtosis-js-lib",
+  "version": "1.0.0",
```
@GoodMorningA1i you'll need to make sure this is updated as part of the `release.sh` script flow

GoodMorningA1i commented on typescript/package.json:
```diff
@@ -1,4 +1,32 @@
 {
+  "name": "kurtosis-js-lib",
+  "version": "1.0.35",
```
Will update this version, just wanted to get an approval on the current implementation

GoodMorningA1i commented:
**Thought-process for Kevin's understanding:**
- The `output.js` I had which was the single .js file that held all compiled typescript files inside of the hidden`typescript/build/` directory is now changed to -> index.js. Since the index.js holds all the dependencies that a user would need, running tsc on index.js automatically compiles all the dependencies as well which is quite neat (this is all of the other compiled .ts files inside of the lib/ directory). Let me know if output.js held a more important purpose that simply acting as the main entry in the package.json so that users can use the package.
- Removed `--module system` from `typescript/scripts/build.sh`, module is to specify the module code generation. I needed it otherwise it would not allow to export type aliases or classes otherwise (like service.ts). However, when I ran `tsc index.ts ...`, it did not require me to use this which is great since I want to keep things default and avoid overcomplicating our typescript codebase more than needed.
- Removed `--moduleResolution node` from `typescript/scripts/build.sh`, moduleResolution is to specify the module resolution strategy. Same reason as `--module system`
- `--declaration true` compiler option on tsc should have gotten rid of imports inside index.d.ts file, but it didn't do so for an unknown reason. Since I was trouble getting things to work with a single compiled .js and .d.ts file, I decided to separate the compilation of all the typescript files into its original files which made it more human-readable and then simply copied over the typescript generated bindings into the hidden build/ folder

**Things to note for myself if we move forward with this:**
- Update the changelog so that it mentions about compiling tsc files separately, not into one single .js file and one single .d.ts file
- test release.sh script to see if it automatically updates the package.json version; would need to confirm with Kevin on if releasing the script in a local branch would be a bad idea (so far, I have tested these commands locally in the shell and via a local bash script)

mieubrisse commented on typescript/scripts/build.sh:
```diff
@@ -13,5 +13,13 @@ lang_root_dirpath="$(dirname "${script_dirpath}")"
+#find_results="$(find ${lang_root_dirpath} -name "*.ts" -not -path "${lang_root_dirpath}/kurtosis_core_rpc_api_bindings/*" -not -path "${lang_root_dirpath}/node_modules/*")"
+
+tsc ${lang_root_dirpath}/index.ts --outDir ${lang_root_dirpath}/build/ --downlevelIteration true --declaration true
+cp -R ${lang_root_dirpath}/kurtosis_core_rpc_api_bindings/ ${lang_root_dirpath}/build/kurtosis_core_rpc_api_bindings/ #TODO - error checking but after confirming if this is okay solution
```
Ooooooo very interesting that this works - so basically we're recreating the directory structure for index.js it looks like. I think this is fine; maybe there's a cleaner way, but whatever.

Only recommendation is DRY, and do`cp -R ${lang_root_dirpath}/kurtosis_core_rpc_api_bindings ${lang_root_dirpath}/` instead. The way you did it is fine, but is vulnerable to typos/changes in the `kurtosis_core_rpc_api_bindings` directory.

The way to think about this is, when you see a trailing `/`, it means "INSIDE of this directory". When you don't see "/", it means "the file/dir itself".

This is also why, if you're moving a bunch of files inside a you expect to exist, MAKE SURE YOU ADD THE TRAILING SLASH! I learned this the hard way in college: I did `mv * typod-directory-name` .... which, because `typod-directory-name` didn't exist, translated to `mv file1 file2 file3 typod-directory-name`.... which meant I ended up with a single file called `typod-directory-name` that contained only the contents of file3. Lost a looooooot of work that day.

Instead, if you do `mv * typod-directory-name/`, the trailing slash tells the shell to put the contents INSIDE the directory. Since `typod-directory-name` doesn't exist and isn't a directory, you'll instead get an error.

mieubrisse replied:
Also: can you add a quick comment explaining why this step is necessary

GoodMorningA1i replied:
This was the code I added to typescript/scripts/build.sh

![image](https://user-images.githubusercontent.com/52847195/130083711-d7e345d7-04bf-4c6c-b5bd-f55dad630029.png)

Compiled index.ts into .js and .d.ts, which then took all of the dependencies and compiled them into their own .js and .d.ts files

And then to get rid of the dependency issue of the typescript generated bindings, I copied those over to the hidden build/ directory where index.js can be found

This got it working when I was locally testing if another repo could register what ContainerRunConfigBuilder.entrypointOverrideArgs is (should be null) which it got correct


![image](https://user-images.githubusercontent.com/52847195/130083778-60900190-9e19-4abc-993d-6eec8adf3582.png)


Here is what test.ts looked like


![image](https://user-images.githubusercontent.com/52847195/130083827-f66a716c-75b7-4b36-b457-46f29e69e3ee.png)

mieubrisse commented on typescript/scripts/build.sh:
```diff
@@ -13,5 +13,13 @@ lang_root_dirpath="$(dirname "${script_dirpath}")"
+tsc ${lang_root_dirpath}/index.ts --outDir ${lang_root_dirpath}/build/ --downlevelIteration true --declaration true
+cp -R ${lang_root_dirpath}/kurtosis_core_rpc_api_bindings/ ${lang_root_dirpath}/build/kurtosis_core_rpc_api_bindings/ #TODO - error checking but after confirming if this is okay solution
+
+#tsc ${lang_root_dirpath}/index.ts --outDir ${lang_root_dirpath}/build --downlevelIteration
```
Get rid of all these commented-out things here and above

mieubrisse reviewed (approved):
Approved with comments

mieubrisse commented:
Minor preference for `tsconfig.json` too

mieubrisse commented on typescript/scripts/build.sh:
```diff
@@ -13,5 +13,13 @@ lang_root_dirpath="$(dirname "${script_dirpath}")"
-tsc ${find_results} --outFile ${lang_root_dirpath}/build/output.js --module system --moduleResolution node --target es2015
+#find_results="$(find ${lang_root_dirpath} -name "*.ts" -not -path "${lang_root_dirpath}/kurtosis_core_rpc_api_bindings/*" -not -path "${lang_root_dirpath}/node_modules/*")"
+
+tsc ${lang_root_dirpath}/index.ts --outDir ${lang_root_dirpath}/build/ --downlevelIteration true --declaration true
```
Minor preference for `tsconfig.json`, but don't worry about it for now - this is fine

mieubrisse commented on typescript/package.json:
```diff
@@ -1,4 +1,32 @@
+  "name": "kurtosis-js-lib",
+  "version": "1.0.35",
+  "description": "This repo contains clients for communicating with the Kurtosis API server, which both testsuites and Kurtosis Modules will use to manipulate the test environment.",
+  "main": "./build/index",
```
Recommend making a specific subdirectory, `build/typescript`, so that you don't inadvertently get cross-language contamination in the future from other languages who dump stuff into `build`
